sympy/polys/matrices/linsolve.py

```diff
@@ -169,41 +168,24 @@ def _linear_eq_to_dict_inner(eqs, syms):
 
 
 def _lin_eq2dict(a, symset):
-    """Efficiently convert a linear equation to a dict of coefficients"""
-    if a in symset:
-        return S.Zero, {a: S.One}
-    elif a.is_Add:
-        terms_list = defaultdict(list)
-        coeff_list = []
-        for ai in a.args:
-            ci, ti = _lin_eq2dict(ai, symset)
-            coeff_list.append(ci)
-            for mij, cij in ti.items():
-                terms_list[mij].append(cij)
-        coeff = Add(*coeff_list)
-        terms = {sym: Add(*coeffs) for sym, coeffs in terms_list.items()}
-        return coeff, terms
-    elif a.is_Mul:
-        terms = terms_coeff = None
-        coeff_list = []
-        for ai in a.args:
-            ci, ti = _lin_eq2dict(ai, symset)
-            if not ti:
-                coeff_list.append(ci)
-            elif terms is None:
-                terms = ti
-                terms_coeff = ci
-            else:
-                raise PolyNonlinearError
-        coeff = Mul(*coeff_list)
-        if terms is None:
-            return coeff, {}
-        else:
-            terms = {sym: coeff * c for sym, c in terms.items()}
-            return  coeff * terms_coeff, terms
-    elif a.is_Equality:
-        return _lin_eq2dict(a.lhs - a.rhs, symset)
-    elif not a.has_free(*symset):
-        return a, {}
-    else:
+    """Efficiently convert a linear equation to a dict of coefficients,
+    returning a tuple (c, d) where c is the symbol-independent
+    expression and d is the dictionary of coefficients with symbols
+    as keys.
+
+    Examples
+    ========
+
+    >>> from sympy.polys.matrices.linsolve import _lin_eq2dict
+    >>> from sympy.abc import x, y
+    >>> _lin_eq2dict(x + 2*y + 3, {x, y})
+    (3, {x: 1, y: 2})
+    """
+    from sympy.solvers.solveset import linear_coeffs, NonlinearError
```

Comment:
There shouldn't be an import dependency this way round (polys importing solvers).